include/pingcap/kv/LockResolver.h

```diff
@@ -91,6 +97,79 @@ struct TxnExpireTime
     int64_t value() { return initialized ? txn_expire : 0; }
 };
 
+// AsyncResolveData is data contributed by multiple threads when resolving locks using the async commit protocol. All
+// data should be protected by the mu field.
+struct AsyncResolveData
+{
+    std::mutex mu;
+    uint64_t commit_ts;
+    std::vector<std::string> keys;
+    bool missing_lock;
+
+    explicit AsyncResolveData(uint64_t _commit_ts, bool _missing_lock) : commit_ts(_commit_ts), missing_lock(_missing_lock) {}
+
+    // addKeys adds the keys from locks to data, keeping other fields up to date. start_ts and _commit_ts are for the
+    // transaction being resolved.
+    //
+    // In the async commit protocol when checking locks, we send a list of keys to check and get back a list of locks. There
+    // will be a lock for every key which is locked. If there are fewer locks than keys, then a lock is missing because it
+    // has been committed, rolled back, or was never locked.
+    //
+    // In this function, resp->locks is the list of locks, and expected is the number of keys. AsyncResolveData.missing_lock will be
+    // set to true if the lengths don't match. If the lengths do match, then the locks are added to asyncResolveData.locks
+    // and will need to be resolved by the caller.
+    void addKeys(std::shared_ptr<::kvrpcpb::CheckSecondaryLocksResponse> resp, int expected, uint64_t start_ts, uint64_t _commit_ts)
+    {
+        std::lock_guard l(mu);
+
+        // Check locks to see if any has been commited or rolled back.
+        if (resp->locks_size() < expected)
+        {
+            // A lock is missing - the transaction must either have been rolled back or committed.
+            if (!missing_lock)
+            {
+                if (_commit_ts != 0 && _commit_ts < commit_ts)
+                {
+                    // commitTS == 0 => lock has been rolled back.
+                    throw Exception("commit TS must be greater than or equal to min commit TS: commit ts: " + std::to_string(_commit_ts)
+                            + ", min commit ts: " + std::to_string(commit_ts),
+                        ErrorCodes::UnknownError);
+                }
+                commit_ts = _commit_ts;
+            }
+            missing_lock = true;
+            if (commit_ts != _commit_ts)
```

Comment:
same question. IMO, commit_ts is the max ts that we have met. Why we must check commit_ts = _commit_ts

Reply:
I don't know why they do this check here. But in a valid transaction, the `commit_ts` of all secondary locks should be same. It makes sense.